DOC: Guide to writing how-tos (#17352)
numpy/numpy: doc/source/user/how-to-how-to.rst

.. _how-to-how-to:

##############################################################################
How to write a NumPy how-to
##############################################################################

How-tos get straight to the point -- they

  - answer a focused question, or
  - narrow a broad question into focused questions that the user can
    choose among.

******************************************************************************
A stranger has asked for directions...
******************************************************************************

> **"I need to fill my tank."**

Reviewer: This is a non-sequitur. If a stranger asks for directions, there should be a destination in the question. It is not clear they are in a car, looking for petrol.

******************************************************************************
Give a brief but explicit answer
******************************************************************************

>   - `"Three miles, take a right at Hayseed Road, it's on your left."`

Reviewer: Let's use the metric system, not miles. "Three kilometers, take a right at Hayseed Road, the petrol station is on your left."

Add helpful details for newcomers ("Hayseed Road", even though it's the only
turnoff at three miles). But not irrelevant ones:

  - Don't also give directions from Route 7.

>   - Don't explain why the town has only one gas station.

Reviewer: "gas" is an Americanism. The rest of the world uses petrol.

If the question is broad, narrow and redirect it
******************************************************************************

 **"I want to see the sights"**

The `See the sights` how-to should link to a set of narrower how-tos:

- Find historic buildings
- Find scenic lookouts
- Find the town center

and these might in turn link to still narrower how-tos -- so the town center
page might link to

   - Find the court house
   - Find city hall

By organizing how-tos this way, you not only display the options for people
who need to narrow their question, you also have provided answers for users
who start with narrower questions ("I want to see historic buildings," "Which
way to city hall?").

******************************************************************************
If there are many steps, break them up
******************************************************************************

If a how-to has many steps:

  - Consider breaking a step out into an individual how-to and linking to it.
  - Include subheadings. They help readers grasp what's coming and return
    where they left off.

******************************************************************************

> Why write how-tos when there's Stack Overflow, Reddit, Gitter...?

Reviewer: Maybe condition their existence on the premise that we have an authoritative answer, something like "we will only write how-tos if we have authoritative answers"

Author: Could you clarify? I'm not sure I understand what's wrong or what the proposed rewording is.

> Why write how-tos when there's Stack Overflow, Reddit, Gitter...?

Reviewer: Something like
```suggestion
When to write a how-to when there's Stack Overflow, Reddit, Gitter...?

```

Author: OK, but I'm still unclear what the objection is to the original heading. I'd change it regardless, but 
* the remaining three headings would not match the new heading because they talk about the values of how-tos rather than conditioning them
* our answers are always authoritative, so the first bullet won't offer clear guidance

Reviewer: ok, thanks.

******************************************************************************

 - We have authoritative answers.
 - How-tos make the site less forbidding to non-experts.
 - How-tos bring people into the site and help them discover other information
   that's here .
 - Creating how-tos helps us see NumPy usability through new eyes.

******************************************************************************
Aren't how-tos and tutorials the same thing?
******************************************************************************

People use the terms "how-to" and "tutorial" interchangeably, but we draw a

> distinction, following `Daniele Procida <https://documentation.divio.com/>`_.

Reviewer: ```suggestion
distinction, following `Daniele Procida's taxonomy of documetation <https://documentation.divio.com/>`_.

```
